src/daft-core/src/array/growable/nested_growable.rs

```diff
@@ -198,3 +198,97 @@ impl<'a> Growable for StructGrowable<'a> {
+        name: String,
+        dtype: &DataType,
+        arrays: Vec<&'a ListArray>,
+        use_validity: bool,
+        capacity: usize,
+        child_capacity: usize,
+    ) -> Self {
+        match dtype {
+            DataType::List(child_field) => {
+                with_match_daft_types!(&child_field.dtype, |$T| {
+                    let child_growable = <<$T as DaftDataType>::ArrayType as GrowableArray>::make_growable(
+                        child_field.name.clone(),
+                        &child_field.dtype,
+                        arrays.iter().map(|a| a.flat_child.downcast::<<$T as DaftDataType>::ArrayType>().unwrap()).collect::<Vec<_>>(),
+                        use_validity,
+                        child_capacity,
+                    );
+                    let growable_validity = ArrowBitmapGrowable::new(
+                        arrays.iter().map(|a| a.validity()).collect(),
+                        capacity,
+                    );
+                    let growable_offsets: Vec<i64> = vec![0];
+                    let child_arrays_offsets = arrays.iter().map(|arr| arr.offsets()).collect::<Vec<_>>();
+                    Self {
+                        name,
+                        dtype: dtype.clone(),
+                        child_growable: Box::new(child_growable),
+                        child_arrays_offsets,
+                        growable_validity,
+                        growable_offsets,
+                    }
+                })
+            }
+            _ => panic!("Cannot create ListGrowable from dtype: {}", dtype),
+        }
+    }
+}
+
+impl<'a> Growable for ListGrowable<'a> {
+    fn extend(&mut self, index: usize, start: usize, len: usize) {
+        let offsets = self.child_arrays_offsets.get(index).unwrap();
+        let offset_slice = &offsets.as_slice()[start..(start + len + 1)];
+        self.child_growable.extend(
+            index,
+            *offset_slice.first().unwrap() as usize,
+            (offset_slice.last().unwrap() - offset_slice.first().unwrap()) as usize,
+        );
+        self.growable_validity.extend(index, start, len);
+
+        let element_lengths = offset_slice.windows(2).map(|window| window[1] - window[0]);
+        for element_length in element_lengths {
+            let last_offset = self.growable_offsets.last().unwrap();
+            self.growable_offsets.push(last_offset + element_length);
+        }
+    }
+
+    fn add_nulls(&mut self, additional: usize) {
+        self.growable_validity.add_nulls(additional);
+        let last_offset = *self.growable_offsets.last().unwrap();
+        self.growable_offsets
+            .extend(repeat(last_offset).take(additional));
+    }
+
+    fn build(&mut self) -> DaftResult<Series> {
+        // Swap out values so we can use them and move it
+        let mut grown_offsets = vec![];
+        let mut grown_validity = ArrowBitmapGrowable::new(vec![], 0);
+        swap(&mut self.growable_validity, &mut grown_validity);
```

Comment:
no need to make your own defaults here. you can use mem::take instead.
https://doc.rust-lang.org/std/mem/fn.take.html

Reply:
Nice, thanks!